okhttp/src/test/java/okhttp3/CacheLockTest.kt

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (C) 2024 Block, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3
+
+import assertk.assertThat
+import assertk.assertions.isEqualTo
+import java.nio.file.Path
+import okio.Closeable
+import okio.FileSystem
+import okio.Path.Companion.toOkioPath
+import org.junit.jupiter.api.AfterEach
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.assertThrows
+import org.junit.jupiter.api.io.TempDir
+
+class CacheLockTest {
+  private lateinit var tempDir: okio.Path
+  private val toClose = mutableListOf<Closeable>()
+
+  @BeforeEach
+  fun setup(
+    @TempDir tempDir: Path,
+  ) {
+    this.tempDir = tempDir.toOkioPath()
+  }
+
+  @AfterEach
+  fun cleanup() {
+    toClose.forEach {
+      it.close()
+    }
+  }
+
+  @Test
+  fun testCacheLock() {
+    openCache(tempDir)
+
+    val ioe =
+      assertThrows<IllegalStateException> {
+        openCache(tempDir)
+      }
+    assertThat(ioe.message).isEqualTo("Cache already open at '$tempDir' in same process")
+  }
+
+  @Test
+  fun testCacheLockAfterClose() {
+    val cache1 = openCache(tempDir)
+
+    cache1.close()
+
+    openCache(tempDir)
+  }
+
+  @Test
+  fun testCacheLockDifferentPath() {
+    openCache(tempDir / "a")
+
+    openCache(tempDir / "b")
+  }
+
+  private fun openCache(directory: okio.Path): Cache {
+    return Cache(directory, 10_000, FileSystem.SYSTEM).apply {
+      // force early LRU initialisation
```

Comment:
I’m a bit torn on this. You don’t get a failure immediately; you get it eventually.

This is probably fine; this new feature is advisory.

Reply:
Yeah, this currently works until it doesn't. Then it ends up as 40+ bugs raised.

Reply:
We can't really do IO operations on the main thread, so this should always be the case.

Which reminds me, for SSL init on main thread https://github.com/square/okhttp/pull/8248

Reply:
I agree with ‘We can’t really do I/O operations on the main thread’, but I don’t agree that OkHttpClient is being initialized on the main thread. We’re an I/O library!